### ISSUE TITLE
Removal of default constructors when decompiling jar breaks method references

### ISSUE THREAD

**Issue by Marcono1234** [leibnitz27/cfr#64]
# CFR version
0.149-SNAPSHOT (commit 2aefade)

# Description
When a jar file is decompiled all class files share one DCCommonState and therefore the same class file cache. This becomes problematic when `CodeAnalyserWholeClass.tryRemoveConstructor(ClassFile)` removes an empty (default) constructor, but a method reference uses it. Restructuring the method reference will then fail [here](https://github.com/leibnitz27/cfr/blob/2aefadeb7d0076e0b1c8e338d97e62bc706b6a09/src/org/benf/cfr/reader/bytecode/analysis/opgraph/op4rewriters/LambdaRewriter.java#L221) and it will emit the raw `LambdaMetafactory` call:
```
public class Main {
    void use(Supplier<?> supplier) {
    }

    void test() {
        this.use((Supplier<Object>)LambdaMetafactory.metafactory(null, null, null, ()Ljava/lang/Object;, <init>(), ()Ljava/lang/Object;)());
    }
}
```

Interestingly it can occasionally successfully restructure the method reference. This apparently happens when the class whose constructor is referenced has not been loaded by the Driver yet (due to random iteration order of jar entries?), but instead is loaded by the class cache. So if there are more differences between `Driver#doJarVersionTypes` and `DCCommonState#loadClassFileAtPath`, there might be more randomly occuring issues.

# Reproduction steps
1. Download [default-constructors.jar.zip](https://github.com/leibnitz27/cfr/files/3935438/default-constructors.jar.zip) (remove `.zip`, this is only to make GitHub happy)
2. Decompile it

**Comment by Col-E** [comment]
Just curious, how do you keep finding all these good issues?

**Comment by Marcono1234** [comment]
For some I thought about tricky corner cases which are easy to forget. However, I noticed most of them while decompiling Minecraft. Now that Mojang publishes the official deobfuscation mappings I started remapping the jars (though not using CFR anymore) and storing the changes between decompiled versions using Git. This made it pretty easy to see these randomly changing things. Due to the rather large code base you also get a great variety of code patterns.
